tests/issues.rs

```diff
@@ -0,0 +1,153 @@
+#![cfg(all(feature = "tokio", feature = "zstd"))]
+
+use std::{
+    io,
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
+
+use async_compression::tokio::write::ZstdEncoder;
+use tokio::io::{AsyncWrite, AsyncWriteExt as _};
+use tracing_subscriber::fmt::format::FmtSpan;
+
+/// This issue covers our state machine being invalid when using adapters
+/// like [`tokio_util::codec`].
+///
+/// After the first [`poll_shutdown`] call,
+/// we must expect any number of [`poll_flush`] and [`poll_shutdown`] calls,
+/// until [`poll_shutdown`] returns [`Poll::Ready`],
+/// according to the documentation on [`AsyncWrite`].
+///
+/// <https://github.com/Nullus157/async-compression/issues/246>
+///
+/// [`tokio_util::codec`](https://docs.rs/tokio-util/latest/tokio_util/codec)
+/// [`poll_shutdown`](AsyncWrite::poll_shutdown)
+/// [`poll_flush`](AsyncWrite::poll_flush)
+#[test]
+fn issue_246() {
+    tracing_subscriber::fmt()
+        .without_time()
+        .with_ansi(false)
+        .with_level(false)
+        .with_test_writer()
+        .with_target(false)
+        .with_span_events(FmtSpan::NEW)
+        .init();
+    let mut zstd_encoder = Wrapper::new(Trace::new(ZstdEncoder::new(DelayedShutdown::default())));
+    futures::executor::block_on(zstd_encoder.shutdown()).unwrap();
```

Comment:
I think we should test `poll_shutdown` by keeping track of numbers of calls to underlying `poll_flush`?

Reply:
I'd want the `poll_flush` to be tested, by making sure it is called if and only if `poll_shutdown` is called.

Reply:
Bit busy now, but in general, I think we want something like this:

```
struct DummyWriter {
    is_poll_flush_called: boolean,
    is_poll_shutdown_called: usize,
}

impl AsyncWrite {
    fn poll_flush(...) {
        assert!(!self.is_poll_shutdown_called);
        self.is_poll_flush_called = true;
        Ready(Ok())
    }

    fn poll_shutdown(...) {
        self.is_poll_shutdown_called = true;
        Ready(Ok())
    }
}
```

And then after `shutdown` is called, we check that both is set to `true` to verify this fix.